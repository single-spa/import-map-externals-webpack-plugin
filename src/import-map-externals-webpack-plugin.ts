import { Compiler, ResolveData, ExternalModule } from "webpack";

export class ImportMapExternalsPlugin {
  declare options: Options;

  constructor(options) {
    this.options = options;
  }

  apply(compiler: Compiler) {
    const importMapPromise: Promise<ImportMap> = this.fetchImportMap();

    compiler.hooks.normalModuleFactory.tap(
      "ImportMapExternalsPlugin",
      (nmf) => {
        nmf.hooks.factorize.tapPromise(
          "ImportMapExternalsPlugin",
          async (resolveData: ResolveData) => {
            const importMap = await importMapPromise;

            if (
              Object.keys(importMap?.imports || {}).includes(
                resolveData.request,
              )
            ) {
              return new ExternalModule(
                resolveData.request,
                compiler.options.output.library?.type || "module",
                resolveData.createData.userRequest || resolveData.request,
              );
            }
          },
        );
      },
    );
  }

  async fetchImportMap(): Promise<ImportMap> {
    const r = await fetch(this.options.importMapUrl);

    if (r.ok) {
      return r.json() as ImportMap;
    } else {
      throw Error(
        `Failed to load import map from url '${this.options.importMapUrl}'`,
      );
    }
  }
}

interface Options {
  importMapUrl: string;
}

interface ImportMap {
  imports?: ModuleMap;
  scopes?: Record<string, ModuleMap>;
}

type ModuleMap = Record<string, string>;
