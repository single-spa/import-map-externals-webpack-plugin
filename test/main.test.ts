import { describe, it, xit } from "@jest/globals";
import webpack from "webpack";
import { ImportMapExternalsPlugin } from "../src/import-map-externals-webpack-plugin";

describe(`ImportMapExternalsPlugin`, () => {
  xit(`externalizes a module`, (done) => {
    webpack(
      {
        mode: "development",
        entry: "./test/fixtures/main.js",
        output: {
          library: {
            type: "module",
          },
        },
        plugins: [
          new ImportMapExternalsPlugin({
            importMapUrl: "https://react.microfrontends.app/importmap.json",
          }),
        ],
        experiments: {
          outputModule: true,
        },
      },
      (err, stats) => {
        console.log(stats?.toString());
        if (err || stats?.hasErrors()) {
          done(`Error during webpack compilation`);
        } else {
          done();
        }
      },
    );
  });
});
