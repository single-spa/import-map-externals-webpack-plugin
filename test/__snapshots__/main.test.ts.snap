// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportMapExternalsPlugin externalizes a module 1`] = `
"import * as __WEBPACK_EXTERNAL_MODULE_single_spa_e1524a87__ from "single-spa";
/******/ var __webpack_modules__ = ({

/***/ "single-spa":
/*!*****************************!*\\
  !*** external "single-spa" ***!
  \\*****************************/
/***/ ((module) => {

module.exports = __WEBPACK_EXTERNAL_MODULE_single_spa_e1524a87__;

/***/ })

/******/ });
/************************************************************************/
/******/ // The module cache
/******/ var __webpack_module_cache__ = {};
/******/ 
/******/ // The require function
/******/ function __webpack_require__(moduleId) {
/******/ 	// Check if module is in cache
/******/ 	var cachedModule = __webpack_module_cache__[moduleId];
/******/ 	if (cachedModule !== undefined) {
/******/ 		return cachedModule.exports;
/******/ 	}
/******/ 	// Create a new module (and put it into the cache)
/******/ 	var module = __webpack_module_cache__[moduleId] = {
/******/ 		// no module.id needed
/******/ 		// no module.loaded needed
/******/ 		exports: {}
/******/ 	};
/******/ 
/******/ 	// Execute the module function
/******/ 	__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/ 
/******/ 	// Return the exports of the module
/******/ 	return module.exports;
/******/ }
/******/ 
/************************************************************************/
/******/ /* webpack/runtime/make namespace object */
/******/ (() => {
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = (exports) => {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/ })();
/******/ 
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
/*!*******************************!*\\
  !*** ./test/fixtures/main.js ***!
  \\*******************************/
__webpack_require__.r(__webpack_exports__);
/* harmony import */ var single_spa__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! single-spa */ "single-spa");


(0,single_spa__WEBPACK_IMPORTED_MODULE_0__.registerApplication)({
  name: "app1",
  activeWhen: ["/"],
  app: {
    async bootstrap() {},
    async mount() {},
    async unmount() {},
  },
});

})();

"
`;
